### ISSUE TITLE
Information about (traditional) DEx prices may not be accurate

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#188]
To track the inline TODO and discussion of https://github.com/OmniLayer/omnicore/pull/173/files#r36941720:

When using `omni_gettransaction`, then the price/ratio returned for traditional DEx sell offers may not be accurate:

---

Alice has only 50 MSC, but offers 1000 MSC for 1000 BTC. The (aweful) way how it's currently handled: the transaction is valid, and [the offer is adjusted](https://github.com/dexX7/bitcoin/blob/d761109327539abfd44af5e050f091ae32662db8/src/omnicore/dex.cpp#L131-L147) to 50 MSC for 50 BTC. The adjusted MSC amount is [fetched from the DB](https://github.com/dexX7/bitcoin/blob/0356f8f8399b6feac7d5e0d2fe7d62ee56016972/src/omnicore/rpctxobject.cpp#L264-L265), but the BTC amount isn't updated:

```java
omni_getbalance moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY 1
{
  "balance" : "50.00000000",  // <----
  "reserved" : "0.00000000"
}
```
```java
// DEx Offer of 1000 MSC for 1000 BTC
omni_sendrawtx "moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY" "0001001400000001000000174876e800000000174876e8000a000000000098968001"
f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec
```
```java
setgenerate true 1
[
  "5a52c577e14918128425d182174cd6b36eaa17d1b92b68604145c5874476cedd"
]
```
```java
omni_gettransaction f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec
{
  "txid": "f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec",
  "fee": "0.00010000",
  "sendingaddress": "moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY",
  "ismine": true,
  "version": 1,
  "type_int": 20,
  "type": "DEx Sell Offer",
  "propertyid": 1,
  "divisible": true,
  "amount": "50.00000000",  // <----
  "bitcoindesired": "1000.00000000",  // <---- !!
  "timelimit": 10,
  "feerequired": "0.10000000",
  "action": "new",
  "valid": true,
  "blockhash": "5a52c577e14918128425d182174cd6b36eaa17d1b92b68604145c5874476cedd",
  "blocktime": 1439467162,
  "block": 103,
  "confirmations": 1
}
```

---

As *temporary* solution (until we handle it better) I suggest to access and reuse a function to recalculate the amounts. This isn't a helper for the RPC layer, but already part of the core logic: [adjustDExOffer](https://github.com/OmniLayer/omnicore/blob/d761109327539abfd44af5e050f091ae32662db8/src/omnicore/dex.cpp#L94). We need to double-check [omni_getactivedexsells](https://github.com/OmniLayer/omnicore/blob/dd7efd7162f6dddaa04ea269c27c9dbb8b229cc4/src/omnicore/rpc.cpp#L1132-L1134) as well, which has a similar inline note.

**Comment by zathras-crypto** [comment]
> As temporary solution

Since DEx v1 doesn't get much traffic, and we're going to issue several consensus breaking changes over the coming months, I advocate for simply changing the behaviour to invalidate an attempt to sell more than you have (basically remove the whole adjustment concept entirely).

I know I'm preaching to the choir here but IMO a sell offer is an explicit action.  If explicit actions cannot be facilitated for whatever reason the transaction is invalid.  I strongly dislike the idea of guessing what the user wanted to do - for example:

* User has 50 MSC in Address A
* User has 100 MSC in Address B
* User has intent to sell 100 MSC from Address B
* User makes following mistake:
  * User sends "Sell 100 MSC" from Address A
  * This explicit action is not possible
  * We *guess* an alternate action that is possible, in this case we assume User wanted to sell 50 MSC not 100 and list the sell accordingly.
* User now has a sell of 50 MSC from Address A, which was not the intent, nor was it the action specified in the message.

TL:DR; we should never modify a message containing an explicit directive to carry out an action.  That we did for DEx was a mistake, and I believe we should rectify.

**Comment by dexX7** [comment]
Yes, I completely agree, and I don't like it for serveral reasons: a notable one, in my opinion, is that it requires even more knowledge about the state, and imho we should rather try to reduce the state dependencies (of course we need to know, whether the sender has sufficient balance to send anyway, but this case is a good example: unconfirmed transactions are even further away from reality here due to the amount adjusting).

*Ideally* transactions are self-sustaining as far as possible, i.e. "one only needs to look at the transaction to learn something about their meaning". In our context this is only possible to some degree, but another example is the whole input selection stuff, and I discussed it with @msgilligan recently, who wanted to build a SPV client that can decode transactions (not necessarily verify them), though one very nasty issue there: to decode class B, we must know who sent the transaction, but to determine that, we need not only the scripts of all inputs, but also their values etc.. Retrieving arbitrary transactions via SPV isn't possible though. I'm probably drifting into off-topic here, but it's pretty similar imho.

I created #189 to discuss the potential consensus change.

---

Nevertheless, regarding the initial issue: we still need

- to address it immediately, in one way or another,
- and probably have to deal with it in the future, too, at least for historical transactions.

While it's not awesome, we could retrieve the data for historical transactions from the DB, given that the transactions are then confirmed (and thus also persisted).

After #189 we can handle unconfrimed transactions, which because the amounts would no longer be adjusted at all.

So it probably comes down to the question, how exactly we tackle those points.
